Forge.TreeWalker/src/TreeWalkerParameters.cs

```diff
@@ -81,6 +81,39 @@ public class TreeWalkerParameters
         /// </summary>
         public Dictionary<string, Func<string, CancellationToken, Task<object>>> ExternalExecutors { get; set; }
 
+        /// <summary>
+        /// The delegate called by Forge when executing SubroutineActions to get an initialized TreeWalkerSession that Forge will call WalkTree upon.
+        /// The TreeWalkerSession should be set up according to the input to execute the Subroutine session.
+        /// Typically this delegate is expected to:
+        ///   1. Create new TreeWalkerParameters with passed in TreeName and TreeInput.
+        ///   2. Update the JsonSchema that maps from the passed in TreeName.
+        ///   3. Create a new ForgeState object. It is important for Subroutine sessions to have their own ForgeState object since the private keyPrefix gets updated.
+        ///   4. Use the same RootSessionId.
+        ///   5. Other parameters can stay the same or be changed. For example, the App could use a different UserContext or ForgeActionsAssembly for different ForgeTrees if they desire.
+        /// </summary>
+        public Func<SubroutineInput, Guid, TreeWalkerParameters, TreeWalkerSession> InitializeSubroutineTree { get; set; }
+
+        /// <summary>
+        /// The unique identifier of the root/parent session that gets passed on to Subroutine sessions.
+        /// RootSessionId will get set to SessionId if not initialized.
+        /// </summary>
+        public Guid RootSessionId { get; set; }
+
+        /// <summary>
+        /// The name of the ForgeTree in the JsonSchema.
+        /// For Subroutines, this is evaluated from the SubroutineInput on the schema.
```

Comment:
>SubroutineInput on the schema [](start = 56, length = 29)

we probably need a Unittest to make sure the Subroutine's input cannot have "RootTree" as the tree name

Reply:
A Subroutine can use the same ForgeTree as a parent session. If the application can prevent that with UTs if they wish.

---
In reply to: [356343697](https://github.com/microsoft/Forge/pull/6#discussion_r356343697) [](ancestors = 356343697)